internal/handlers/initrd_addrsize.go

```diff
@@ -0,0 +1,92 @@
+package handlers
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/go-chi/chi/v5"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/openshift/assisted-image-service/pkg/imagestore"
+	"github.com/openshift/assisted-image-service/pkg/isoeditor"
+	"github.com/openshift/assisted-image-service/pkg/overlay"
+)
+
+const initrdAddrsizePathInISO = "images/initrd.addrsize"
+
+type initrdAddrSizeHandler struct {
+	ImageStore imagestore.ImageStore
+	client     *AssistedServiceClient
+}
+
+var _ http.Handler = &initrdAddrSizeHandler{}
+
+func (h *initrdAddrSizeHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	imageID := chi.URLParam(r, "image_id")
+
+	version := r.URL.Query().Get("version")
+	if version == "" {
+		httpErrorf(w, http.StatusBadRequest, "'version' parameter required for initrd download")
+		return
+	}
+
+	arch := r.URL.Query().Get("arch")
+	if arch != "s390x" {
+		httpErrorf(w, http.StatusBadRequest, "'initrd.addrsize' file only available for s390x architecture")
+		return
+	}
+
+	isoPath := h.ImageStore.PathForParams(imagestore.ImageTypeFull, version, arch)
+
+	initrdReader, lastModified, code, err := initrdOverlayReader(h.ImageStore, h.client, r)
+	if err != nil {
+		httpErrorf(w, code, err.Error())
+		return
+	}
+	// reader need to be closed
+	defer initrdReader.Close()
+
+	fileName := fmt.Sprintf("%s-initrd.addrsize", imageID)
+	addrsizeReader, err := isoeditor.GetFileFromISO(isoPath, initrdAddrsizePathInISO)
+	if err != nil {
+		log.Errorf("Error retrieving initrd.addsize file: %v, isoPath; %s\n", err, isoPath)
+		httpErrorf(w, http.StatusInternalServerError, "Failed to get initrd: %v", err)
+		return
+	}
+
+	// get the size of the initrd including the embedded ignition
+	sizeOfInitrd, err := initrdReader.Seek(0, io.SeekEnd)
+	if err != nil {
+		httpErrorf(w, http.StatusInternalServerError, "Failed to determine size of initrd: %v", err)
+		return
+	}
+
+	buf := new(bytes.Buffer)
+	binary.Write(buf, binary.BigEndian, sizeOfInitrd)
+
+	// create overlay reader for initrdsize
+	overlayAddrsize := overlay.Overlay{
+		Reader: bytes.NewReader(buf.Bytes()),
+		Offset: 8,
+		Length: 8,
+	}
+
+	aReader, err := overlay.NewOverlayReader(addrsizeReader, overlayAddrsize)
+	if err != nil {
+		httpErrorf(w, http.StatusInternalServerError, "Failed to create overlay reader for initrd.addrsize: %v", err)
+		return
+	}
+
+	modTime, err := http.ParseTime(lastModified)
+	if err != nil {
+		log.Warnf("Error parsing last modified time %s: %v", lastModified, err)
+		modTime = time.Now()
+	}
```

Comment:
Do we really need this overlaay? Can't we just serve the addrsize bytes buffer?

Reply:
Yes ... because I need to take the initrd.addrsize from file and copy the size of initrd appended with the ignition into the last 8 byte and give return the new initrd.addrsize file back.

Reply:
Ah I see. Isn't that just 8 bytes? Can't we just read them and append the size and serve the 16-byte buffer?

Reply:
I'm not sure how that addresses my question. Why do a "live edit" of an existing file, when it's just 16 bytes? Read the first 8 bytes (the psw?), append the size of the new initrd, and serve that. The overlay is useful for editing huge files, it's over-engineered for 2 numbers

Reply:
Why is this conversation marked as resolved?

Reply:
Sorry ... closed accidentally. Overlay removed and return new buffer.